### ISSUE TITLE
Proposing using GPS time instead of epoch

### ISSUE THREAD

**Issue by lasdasdas** [UnaNancyOwen/VelodyneCapture#17]
Hey there again, 

So, as we continue to use your drivers, we are coming up with new ideas. As you probably already know, a very common use case of the Velodyne sensors is mobile mapping . When using several sensors, synchronization is important, and using the GPS timestamp is often the most precise way to do so. Both the VLP16 and the VLP32 happen to have this feature(The Velodyne sensors return the microseconds since the last hour). I would love to see this implemented, and I would also like to help develop it. 

The timestamp was implemented a year ago or so, but it uses unix epoch timestamp, whilst ignoring the data packet timestamp. It is true that when the sensor is working "live" with asio, the time might me consistent and useful, however, when opening PCAPS, knowing the timestamp is crucial to a proper post processing synchronization. 

### What we propose.

- Putting this new option under a precompiler option to switch between unix and GPS time, UNIX by default in order to allow older code to have same behavior.  
- Initially signing each point with the data packet GPS timestamp and then maybe implement a look up table to add each laser shoot their proper delay as specified on the corresponding user manual. 

If you think it might be worth it, we could open a branch and work there. If not, I will probably fork and implement it there, but your feedback would be much appreciated. 

Cheers!

**Comment by UnaNancyOwen** [comment]
I agree with your proposal.
I think GPS time-stamp is enable when connected GPS receiver module (, but I might be wrong).
If so, It should be use GPS time-stamp when it is connected. (If it is not connected, It should be use UNIX time-stamp.)

However, I don't have Velodyne sensor and GPS receiver module.
If you can develop this feature, I welcome your contribution.
You can choose your preferred contribution method (send pull request or fork this project).
If you want commit right, I will give it.

**Comment by lasdasdas** [comment]
Hey, sorry, I have been away,

Okey, AFAIK the data packets are always stamped with the microseconds since the last hour. If the Velodyne doesen't have a GPS module, the microseconds are counted by a RTC considering the "start of the new hour" the boot time of the sensor. Then, if a GPS signal is received, the RTC leaps to the GPS time. 

In conclusion,  I would rather let the user decide between the two clocks, with a precompiler variable or a flag in the class, rather than to do it manually by listening on the GPS port. The timestamp is well behaved in both cases, and the use case is also different. 

I have checked the class, and all of this should be pretty easy to implement. Velodyne also offers "timestamp" to point correction using the look up tables for shooting delays. These delays can also be used for "pointwise" azimuth interpolation in the future. 

### Let's go

Alright, for now, I think it is best to simply put the data packet timestamp to every single point, without accounting for delays (same approach as the Unix time). And enable the option *on runtime* with a class function. Very simple, I will test it with a VLP16, and some PCAPS. If you open a branch in this repo, I will do it here, I'd rather do that. 
```
void setUnixTime(bool switch){
//Default is unix time // false is GPS
... 
}
```
Later on, we can think of the pointwise interpolation(timestamp and azimuth ), given the delay look up table of Velodyne.

**Comment by UnaNancyOwen** [comment]
LGTM 👍  It appears that I should give you the commit right. Just a moment.

**Comment by lasdasdas** [comment]
I will close this, as it was just merged in 5ce11453ec4202826b1d557eb3fe52f6b8611e82.

Cheers!
